compiler/hash-tir/src/node.rs

```diff
@@ -0,0 +1,111 @@
+//! Defines the type wrapper for TIR nodes, which contains some
+//! metadata about the node, and the data itself.
+//!
+//! The metadata includes the node origin, which points to the
+//! AST node that the TIR node was generated from, if any.
+//!
+//! Nodes normally live in stores, which can be created through the
+//! `crate::environment::stores::tir_node_*` macros.
+use derive_more::{Deref, DerefMut};
+use hash_ast::ast::AstNodeId;
+use hash_source::{location::Span, SourceId};
+use hash_storage::store::statics::SingleStoreValue;
+
+use crate::ast_info::HasNodeId;
+
+/// Represents a node in the TIR.
+///
+/// Each node has an origin, and data associated with it.
+#[derive(Debug, Deref, DerefMut, Copy, Clone, PartialEq, Eq)]
+pub struct Node<Data> {
+    pub origin: NodeOrigin,
+    #[deref]
+    #[deref_mut]
+    pub data: Data,
+}
+
+/// Helper implementation for `Node`s which are also `SingleStoreValue`s,
+/// so that a node and its entry in the store can be created in one go.
+impl<Data> Node<Data>
+where
+    Self: SingleStoreValue,
+{
+    /// Create a new node with the given data and origin, and insert it into the
+    /// store.
+    pub fn create_at(data: Data, origin: NodeOrigin) -> <Self as SingleStoreValue>::Id {
+        Self::create(Self::at(data, origin))
+    }
+
+    /// Create a new node with the given data and a generated origin, and insert
+    /// it into the store.
+    pub fn create_gen(data: Data) -> <Self as SingleStoreValue>::Id {
+        Self::create(Self::gen(data))
+    }
+}
+
+impl<Data> Node<Data> {
+    /// Create a new node with the given data and origin.
+    pub fn at(data: Data, origin: NodeOrigin) -> Self {
+        Self { data, origin }
+    }
+
+    /// Create a new node with the given data and a generated origin.
+    pub fn gen(data: Data) -> Self {
+        Self { data, origin: NodeOrigin::Generated }
+    }
+
+    /// Get the node ID of this node.
+    pub fn node(&self) -> Option<AstNodeId> {
```

Comment:
I wonder if we can get away with a node always having a `AstNodeId`

Reply:
I think sometimes this is not possible, for example the primitives are made up of nodes but are not defined in any source file.

Reply:
Also, is there a need to keep track of what is inferred or generated?

Reply:
Yes, because it can lead to better error messages. For example, "type X given here` vs "type X inferred from this expression"